rosidl_generator_cpp/resource/msg__struct.hpp.em

```diff
@@ -81,55 +82,48 @@ struct @(spec.base_type.type)_
 {
   using Type = @(spec.base_type.type)_<ContainerAllocator>;
 
-@# constructors (with and without allocator)
-@[for (alloc_type, alloc_name) in [['', ''], ['const ContainerAllocator & ', '_alloc']]]@
-  @('explicit ' if alloc_type else '')@(spec.base_type.type)_(@(alloc_type + alloc_name))
-@# generate initializer lists
-@[  if alloc_name == '_alloc']@
-// *INDENT-OFF* (prevent uncrustify from making unnecessary indents here)
-@[  end if]@
+  explicit @(spec.base_type.type)_(rosidl_generator_cpp::MessageInitialization init = rosidl_generator_cpp::MessageInitialization::ALL)
+  {
+    if (init == rosidl_generator_cpp::MessageInitialization::ALL) {
+      this->initialize_all();
+    } else if (init == rosidl_generator_cpp::MessageInitialization::SKIP) {
+      // no work to do
+    } else if (init == rosidl_generator_cpp::MessageInitialization::ZERO) {
+      this->zero_initialize();
+    } else if (init == rosidl_generator_cpp::MessageInitialization::DEFAULTS_ONLY) {
+      this->initialize_defaults_only();
+    }
+  }
+
 @{
-used_alloc_name = False
-leading_colon = ':'
+init_list = ":"
+used_alloc = False
 for field in spec.fields:
     # dynamic arrays require allocator if available
-    if alloc_name and field.type.is_array and \
+    if field.type.is_array and \
             (field.type.array_size is None or field.type.is_upper_bound):
-        if leading_colon == ' ':
-            print(',')
-        print('  %s %s(%s)' % (leading_colon, field.name, alloc_name), end='')
-        used_alloc_name = True
-        leading_colon = ' '
-
-    # if default value is available set it in initializer list
-    elif field.type.is_primitive_type() and field.default_value is not None:
-        cpp_value = value_to_cpp(field.type, field.default_value)
-        if leading_colon == ' ':
-            print(',')
-        print('  %s %s(%s)' % (leading_colon, field.name, cpp_value), end='')
-        leading_colon = ' '
-if leading_colon == ' ':
-    print()
+        init_list += ' %s(_alloc),' % (field.name)
+        used_alloc = True
+
+# Always remove the last element; it will either be a trailing comma we do
+# not want, or a colon we do not want.
+init_list = init_list[:-1]
 }@
-@[  if alloc_name == '_alloc']@
-// *INDENT-ON*
-@[  end if]@
+  explicit @(spec.base_type.type)_(const ContainerAllocator & _alloc, rosidl_generator_cpp::MessageInitialization init = rosidl_generator_cpp::MessageInitialization::ALL)@(init_list)
```

Comment:
This line will get pretty long. Together with the comment above this can be written as:

```
explicit ... MessageInitialization::ALL)
@[if init_list]@
    : @(', '.join(init_list))
@[end if]@
```

If the length of that line still bothers us we could even put each initialization on a separate line.

Reply:
Yeah, I actually think I prefer to put each on its own line so that we never run into line length problems.  I'll change the code to do that.